src/sysupdate/sysupdated.c

```diff
@@ -0,0 +1,1916 @@
+        HASHMAP_FOREACH(t, m->targets) {
+                _cleanup_strv_free_ char **target_appstream = NULL;
+                r = target_get_appstream(t, &target_appstream);
+                if (r < 0)
+                        return r;
+
+                r = strv_extend_strv(&urls, target_appstream, true);
+                if (r < 0)
+                        return r;
+        }
+
+        r = sd_bus_message_new_method_return(msg, &reply);
+        if (r < 0)
+                return r;
+
+        r = sd_bus_message_append_strv(reply, urls);
+        if (r < 0)
+                return r;
+
+        return sd_bus_send(NULL, reply, NULL);
+}
+
+static const sd_bus_vtable manager_vtable[] = {
+        SD_BUS_VTABLE_START(0),
+
+        SD_BUS_METHOD_WITH_ARGS("ListTargets",
+                                SD_BUS_NO_ARGS,
+                                SD_BUS_RESULT("a(sso)", targets),
+                                method_list_targets,
+                                SD_BUS_VTABLE_UNPRIVILEGED),
+
+        SD_BUS_METHOD_WITH_ARGS("ListJobs",
+                                SD_BUS_NO_ARGS,
+                                SD_BUS_RESULT("a(tsuo)", jobs),
+                                method_list_jobs,
+                                SD_BUS_VTABLE_UNPRIVILEGED),
+
+        SD_BUS_METHOD_WITH_ARGS("ListAppStream",
+                                SD_BUS_NO_ARGS,
+                                SD_BUS_RESULT("as", urls),
+                                method_list_appstream,
+                                SD_BUS_VTABLE_UNPRIVILEGED),
+
+        SD_BUS_SIGNAL_WITH_ARGS("JobRemoved",
+                                SD_BUS_ARGS("t", id, "o", path, "i", status),
+                                0),
+
+        SD_BUS_VTABLE_END
+};
+
+static const BusObjectImplementation manager_object = {
+        "/org/freedesktop/sysupdate1",
+        "org.freedesktop.sysupdate1.Manager",
+        .vtables = BUS_VTABLES(manager_vtable),
+        .children = BUS_IMPLEMENTATIONS(&job_object, &target_object),
+};
+
+static int manager_add_bus_objects(Manager *m) {
+        int r;
+
+        assert(m);
+
+        r = bus_add_implementation(m->bus, &manager_object, m);
+        if (r < 0)
+                return r;
+
+        r = bus_log_control_api_register(m->bus);
+        if (r < 0)
+                return r;
+
+        r = sd_bus_request_name_async(m->bus, NULL, "org.freedesktop.sysupdate1", 0, NULL, NULL);
+        if (r < 0)
+                return log_error_errno(r, "Failed to request name: %m");
+
+        r = sd_bus_attach_event(m->bus, m->event, 0);
+        if (r < 0)
+                return log_error_errno(r, "Failed to attach bus to event loop: %m");
+
+        return 0;
+}
+
+static bool manager_idle(void *userdata) {
+        Manager *m = ASSERT_PTR(userdata);
+
+        if (!hashmap_isempty(m->jobs))
+                return false;
+
+        return true;
+}
+
+static void manager_check_idle(Manager *m) {
+        if (hashmap_isempty(m->jobs))
```

Comment:
mayber add `assert(m)` first, as usual

also, we like if the primary code paths is the unindented one, and the exceptional case is the indented one.

hence, maybe invert this check here, and do a `return`.

```c
assert(m);

if (!hashmap_isempty(m->jobs))
        return;

hashmap_clear(…);
log_debug(…);
```